### ISSUE TITLE
Sanity tests should listen to prestashop event

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#20359]
#### Is your feature request related to a problem?

We had to release an emergency 1.7.6.7 release because we did not notice early enough that a JS event was not correctly emitted/received by JS modules plugged on Product FO page. Because modules and themes are core to PrestaShop ecosystem, this was a critical issue.

We can expect more similar issues in the future if we do not monitor correctly that
- on backend side, hooks are correctly dispatched and received for critical user paths
- on frontend side, prestashop JS events are correctly dispatched and received for critical user paths

#### Describe the solution you'd like

On frontend side, our browser tests powered by Puppeeteer should listen to prestashop JS events and assert they are correctly dispatched and received for critical user paths. It seems [Puppeeteer provide the necessary tools for it](https://github.com/puppeteer/puppeteer/issues/573).

So I imagine something like this:
```javascript
  // tests/UI/campaigns/sanity/05_cartFO/01_editCheckCart.js
  it('should add product to cart and check that the number of products was updated in cart header', async function () {

    await productPage.exposeFunction('updateCart', (e) => {
      eventLogger.logPSEvent(e);
    });

    await testContext.addContextItem(this, 'testIdentifier', 'addProductToCart2', baseContext);
    await productPage.addProductToTheCart(page);

    await eventLogger.checkJSEventWasTriggered'updateCart');

    // getNumberFromText is used to get the notifications number in the cart
    const notificationsNumber = await homePage.getNumberFromText(page, homePage.cartProductsCount);
    await expect(notificationsNumber).to.be.equal(2);
  });
```

#### Alternatives you've considered

QA analysts could check manually that JS events are fired but this is time-consuming

**Comment by SimonGrn** [comment]
(FYI we're using [Playwright](https://github.com/microsoft/playwright) now, not Puppeteer :sweat_smile: )
We know it's something we need to check, we already have some pieces of the whole system in our linkchecker: it checks every requests and tracks the failed ones (404 and other), and also collect all the errors and warnings in the console natively.
We also use JS injection when we need to execute some selectors checking (if a DOM element is very picky for example, or if we want to get the inner text of an element) on our test campaigns.
We can combine these two pieces to make something, but I would not use our sanity campaign.
This campaign is tailored for a specific usage: checking basic features are working properly. If one day you want to add more stuff to check with listeners (more pages, or features) we will have to modify the scope of the sanity for something not related to its core usage.
I would advise for a new campaign designed to do specifically what you want. That way we can modify it, extend it, whatever you want without modifying the scope of the sanity and its usability. If necessary we could also combine the two by executing them side to side (I guess you're thinking about PR checks ?).
What do you think ?

**Comment by boubkerbribri** [comment]
First step for listeners is done with #20368 : Adding listeners on existing campaigns.
Next step will be : Listen and verify specific event js and for that should be done in a new campaign.

**Comment by matks** [comment]
> First step for listeners is done with #20368 : Adding listeners on existing campaigns.
> Next step will be : Listen and verify specific event js and for that should be done in a new campaign.

There will be a tricky part: the JS events we want to listen to are not global events, they are being sent on global variable `prestashop` that, if I am not wrong, is only available on `document.ready`

You can look at https://github.com/PrestaShop/ps_shoppingcart/blob/master/ps_shoppingcart.js for example:
```
$(document).ready(function () {

  prestashop.on(
    'updateCart',
    function (event) {
      // here our assertion code should be triggered
    )
  );
}
```

So we need
- to wait for `$(document).ready()`
- then start listening on event
- then trigger the right action (example: add product to cart)
- then check event was received

How does it look to you ? Doable ?

**Comment by SimonGrn** [comment]
Playwright has some native ways to wait for the page to be "ready", like waiting for the "domloaded" event, or waiting for all connections to be closed. I'm almost 100% certain we can plug some sort of check for the "document.ready" event if it's not natively handled.
As for executing arbitrary JS code in the page, that shouldn't be a problem.

This is the perfect example as to why this sort of checks should be in its own campaign, as it may become more complex by the time and shouldn't be tied to the sanity campaign.

**Comment by matks** [comment]
@SimonGrn @boubkerbribri Other ideas to start covering hooks behavior into Playwright tests (might be another topic, please tell me if you want me to create a new issue)

## Test display hooks

We could perform the following steps:
1. when prestashop is installed, load and install module ps_qualityassurance
2. configure ps_qualityassurance to be hooked on ALL display Hooks with a specific div structure (see below)
3. in Playwright tests, when opening a page (FO or BO) check the div exists

Example of a structure:
```html
<div id="hook-[hook-name]" style="hidden"></div>
```

This would at least validate that display hooks are called correctly.

In terms of doability:
1. How hard is it, into a Playwright test, to add a step "I install module X" ?
2. This step could be done with an option into ps_qualityassurance. Something like "load fixtures".
3. This one is easy I believe 😁 

## Test action hooks

Action hooks are invisible so you will not be able to parse the DOM to assert they work correctly. What could be done is this:

1. we implement into ps_qualityassurance a "logger" behavior: every time a hook calls the module, it registers the call and the parameters
2. we implement into ps_qualityassurance a small API endpoint that returns the logs as JSON data
3. when prestashop is installed, load and install module ps_qualityassurance
4. when Playwright visits pages, this is going to trigger hooks and this will be logged by ps_qualityassurance
5. you can them implement a step where you ask the endpoint "did hook X call you with parameters Y ?"

This system is a bit hacky (implement a logger, an endpoint, call the endpoint ...) and will be hard to debug. But I got nothing better for now.

**Comment by SimonGrn** [comment]
The first part seems pretty easy, with just a few tricks I guess we can automate all of this.
Second part is as you mentioned yourself, a bit more tricky. A bit. Currently we don't have the time nor the resources to work on this kind of project.
We can try to make your first suggestion with a "nice to have" priority. That's all I can offer at the moment, sadly :(

>How hard is it, into a Playwright test, to add a step "I install module X" ?

Kinda complicated but we can circumvent this by pre-installing the module with some shell action I guess.

**Comment by matks** [comment]
> Second part is as you mentioned yourself, a bit more tricky. A bit. Currently we don't have the time nor the resources to work on this kind of project.

On Playwright side, the things to do are

For "display hooks"

- automate "install module X"
- detect divs exist in DOM

For "action hooks"

- automate "install module X"
- send a HTTP request to a API endpoint provided by ps_qualityassurance and perform assertions against given answer

ps_qualityassurance work is on our side 😉

**Comment by SimonGrn** [comment]
Thing is this is a whole new campaign, and we would like to finish what we already have in the pipe before starting something else. This is an interesting topic, for sure, be we currently have at least 3 interesting topics to explore too :(

**Comment by matks** [comment]
> Thing is this is a whole new campaign

I dont know ... in theory, since prestashop business relies on modules and themes correct behavior (this is where the money come from 😄 ) we could consider than ensuring the most important hooks should be tested into sanity tests ?

Dont go for the big picture immediately 😊  we could do a small POC for the most critical hooks (modok can tell us "the most used hooks in prestashop") and then slowly add more, as we get bug reports or we enhance the tests ?

**Comment by SimonGrn** [comment]
I guess this is a more interesting approach, however I can assure you all the other topics in the pipe are kinda important as well... we're lacking firepower, that's why I'm so desperately looking for a new automation engineer :stuck_out_tongue: 
Automating the installation of modules is something we don't know if we should automate using the UI as it's really brittle, and I wanted to keep it on the manual campaign for the moment.
We have other alternatives, like using a symfony-cli command.

>we could consider than ensuring the most important hooks should be tested into sanity tests ?

Our sanity could use a little love (that's true) but it's really important to balance usefulness, time, and the (short) list of features that should be tested. This is a tricky path. I'm gonna check what we can do.

**Comment by PierreRambaud** [comment]
I'm not in favor of adding this kind of tests in playwright / sanity, this is something that must be done with unit tests.
The main reason is, display & action hooks are already tested because we install PrestaShop with all our native modules:
- https://github.com/PrestaShop/ps_facetedsearch/blob/master/src/Hook/Category.php#L34
- https://github.com/PrestaShop/ps_facetedsearch/blob/master/src/Hook/ProductSearch.php#L35
- https://github.com/PrestaShop/ps_facetedsearch/blob/master/src/Hook/Design.php#L31

We should keep the UI tests as generic as possible like we are already doing it. And if you want to go further, you must implement unit tests to be sure "with this `input` you have this `output`".

**Comment by matks** [comment]
> I'm not in favor of adding this kind of tests in playwright / sanity, this is something that must be done with unit tests.

The recent release of 1.7.6.7 proved that unit tests are not enough. If we had JS unit tests on event being dispatched, they would have been green for 1.7.6.6 even though the JS event would not be catched by modules listening to them.

Unit tests are more accurate and indeed if we could unit test the very logic of hooks it would be good, but we need another way to assert that, in a shop/application context, the hooks are working.

⬇️ 

A Symfony application is unit tested but if the database username is wrong, it will not work 😁  this is what integration/acceptance/e2e tests are there for

IMO some hooks should be part of sanity tests because sanity tests cover the heart of PrestaShop behavior. PrestaShop without modules is nothing. Some hooks, especially the ones for FO, are critical and should be tested too. Other hooks, I agree, can go into another campaign, be it nightly or a dedicated campaign.

✅ 

Basically in this issue I am trying to find a way to ensure a scenario similar to 1.7.6.7 emergency release, 2 days after 1.7.6.6, does not happen again. Right now the coverage of hooks (unit or integration or acceptance) is ... zero 😛  . We can discuss, debate, brainstorm how and where this can be done 😄  I'm open for it, I just want us to achieve a way to be able to release incoming 1.7.7.0 without fearing a similar scenario.

**Comment by SimonGrn** [comment]
Thing is, a hook is a "low level" feature, which can only be correctly tested through a "high level" test (UI).
So... who's responsible for this :stuck_out_tongue_closed_eyes: ?
On a more serious note, I don't think it's **that** useless to test a _few_ hooks through UI tests and put that in the sanity campaign. 
In the long term, we could have a dedicated campaign testing all the hooks, after all it's just a giant loop !

BTW I know unit tests are easier and faster but in the end using UI tests is the only way to make sure the whole stack works as expected.

**Comment by PierreRambaud** [comment]
> The recent release of 1.7.6.7 proved that unit tests are not enough. If we had JS unit tests on event being dispatched, they would have been green for 1.7.6.6 even though the JS event would not be catched by modules listening to them.

We don't have unit tests in the front this is why we had not caught it :sweat_smile: 

> A Symfony application is unit tested but if the database username is wrong, it will not work grin this is what integration/acceptance/e2e tests are there for

Any application must be unit tested then, use integration tests when it's a low level tests.
End to End or UI tests must be as simple as possible and represent what users will do :thinking: 


We must combine test suites (units => integration => UI tests), and currently, in our UI tests, we are already testing hooks with pre-installed modules.  As @SimonGrn said, it's a "low level" feature, that must be tested with unit / integration tests.

**Comment by matks** [comment]
> We don't have unit tests in the front this is why we had not caught it 😅

The issue was inside bundled JS code, not source code. Unit and integration tests do not test the compiled code, do they ?

What is worse: du to the nasty nature of the bug, Core Javascript would receive the event but not 3rd party modules because the JS crash happened later.

> Any application must be unit tested then, use integration tests when it's a low level tests.
> End to End or UI tests must be as simple as possible and represent what users will do 🤔

A user _will_ install custom theme and custom modules and he expects them to work if they are valid. In 1.7.6.6 a valid module plugged on the JS events impacted, although valid, would not work.

> We must combine test suites (units => integration => UI tests), and currently, in our UI tests, we are already testing hooks with pre-installed modules. As @SimonGrn said, it's a "low level" feature, that must be tested with unit / integration tests.

I totally agree with you that we need unit AND integration AND UI tests on this area 😄  . However
- unit tests validate that your logic, at code level, is valid
- integration tests validate that your components, when plugged together, do behave in the expected way
- application/acceptance/UI tests validate that your application, when set up and running, do what it should do when a user uses it

When you arrive on a new codebase with no tests, UI tests have the best short-term ROI as they immediately tell you whether the behavior is correct from a user POV (here: I install a module plugged on hook FooBar, it works) while unit tests help you refactor the code which means they have a better ROI long-term.

We want both short-term and long-term outcomes 😄  but 1.7.7.0 is right behind the corner. This is why I aim for short-term here.

**Comment by PierreRambaud** [comment]
> The issue was inside bundled JS code, not source code. Unit and integration tests do not test the compiled code, do they ?

It's like saying all our integrations tests / unit tests are useless because they don't test cached files. If the problem is inside a bundled JS file, it's also in the source code, otherwise you have no way to fix it :thinking: 

> A user will install custom theme and custom modules and he expects them to work if they are valid. In 1.7.6.6 a valid module plugged on the JS events impacted, although valid, would not work.

This is already something we test with our native modules, we check that all our hooks and features are working with the user experience.

>I totally agree with you that we need unit AND integration AND UI tests on this area smile . However
> 
> * unit tests validate that your logic, at code level, is valid
> * integration tests validate that your components, when plugged together, do behave in the expected way
> * application/acceptance/UI tests validate that your application, when set up and running, do what it should do when a user uses it

Unit tests => this means all possible values, what happened if you have this kind or this kind of value, etc.

> When you arrive on a new codebase with no tests, UI tests have the best short-term ROI as they immediately tell you whether the behavior is correct from a user POV (here: I install a module plugged on hook FooBar, it works) while unit tests help you refactor the code which means they have a better ROI long-term.

With this in mind, you never take the time to do proper unit tests and add a huge complexity and maintainability in UI tests.
Also, with UI tests, you're not fully tested the logic, only if the text (because you never test the HTML is properly rendered) is good which is also why even if all your UI tests are green and if a user type a different character than yours, the interface is failing.

Since we are already testing many hooks because of native modules, you're just going to add complexity to UI tests for something that can be tested in another way :/

**Comment by hibatallahAouadni** [comment]
Hello @SimonGrn @boubkerbribri 

I just noticed that the PR is merged! So shall we close the issue?

**Comment by boubkerbribri** [comment]
> Hello @SimonGrn @boubkerbribri
> 
> I just noticed that the PR is merged! So shall we close the issue?

If you are talking about this PR #20368 , there still some work to be done on Unit tests (metionned in comments) and the UI tests.

**Comment by hibatallahAouadni** [comment]
@boubkerbribri So the status of this issue is **`To Do`** ?

**Comment by matks** [comment]
@hibatallahAouadni Yes. Also this PR was the 1st part for solving this issue, there is still a long path ahead.
